src/parallel/NeighborAcquirer.cpp

```diff
@@ -74,49 +75,52 @@ std::tuple<std::vector<CommunicationPartner>, std::vector<CommunicationPartner>>
-	std::vector<CommunicationPartner> comm_partners02;
+	// the regions I own and want to send: ranks<regions<regionData>>
+	std::vector<std::vector<std::vector<unsigned char>>> sendingList(num_processes);
 
-	i = 0;
-	while (i != num_bytes_receive) {
-		int rank;
-		int regions;
+	std::vector<CommunicationPartner> comm_partners02{};
+	bufferPosition = 0;
+	while (bufferPosition < num_bytes_receive /*== buffer length*/) {
 
-		memcpy(&rank, incomingDesiredRegionsVector.data() + i, sizeof(int));
-		i += sizeof(int);  // 4
-		memcpy(&regions, incomingDesiredRegionsVector.data() + i, sizeof(int));
-		i += sizeof(int);  // 4
+		int rank{};
+		memcpy(&rank, incomingDesiredRegionsVector.data() + bufferPosition, sizeof(int));
+		bufferPosition += sizeof(int);  // 4
+		int regions{};
+		memcpy(&regions, incomingDesiredRegionsVector.data() + bufferPosition, sizeof(int));
+		bufferPosition += sizeof(int);  // 4
 
-		for (int j = 0; j < regions; j++) {
+		for (int regionId = 0; regionId < regions; ++regionId) {
 			HaloRegion unshiftedRegion{};
-			memcpy(unshiftedRegion.rmin, incomingDesiredRegionsVector.data() + i, sizeof(double) * 3);
-			i += sizeof(double) * 3;  // 24
-			memcpy(unshiftedRegion.rmax, incomingDesiredRegionsVector.data() + i, sizeof(double) * 3);
-			i += sizeof(double) * 3;  // 24
-			memcpy(unshiftedRegion.offset, incomingDesiredRegionsVector.data() + i, sizeof(int) * 3);
-			i += sizeof(int) * 3;  // 12
-			memcpy(&unshiftedRegion.width, incomingDesiredRegionsVector.data() + i, sizeof(double));
-			i += sizeof(double);  // 4
+			memcpy(unshiftedRegion.rmin, incomingDesiredRegionsVector.data() + bufferPosition, sizeof(double) * 3);
+			bufferPosition += sizeof(double) * 3;  // 24
+			memcpy(unshiftedRegion.rmax, incomingDesiredRegionsVector.data() + bufferPosition, sizeof(double) * 3);
+			bufferPosition += sizeof(double) * 3;  // 24
+			memcpy(unshiftedRegion.offset, incomingDesiredRegionsVector.data() + bufferPosition, sizeof(int) * 3);
+			bufferPosition += sizeof(int) * 3;  // 12
+			memcpy(&unshiftedRegion.width, incomingDesiredRegionsVector.data() + bufferPosition, sizeof(double));
+			bufferPosition += sizeof(double);  // 4
 
 			// msg format one region: rmin | rmax | offset | width | shift
-			auto shiftedRegionShiftPair = getPotentiallyShiftedRegions(globalDomainLength, unshiftedRegion);
-
-			std::vector<HaloRegion> regionsToTest = shiftedRegionShiftPair.first;
-			std::vector<std::array<double, 3>> shifts  = shiftedRegionShiftPair.second;
-
+			const auto [regionsToTest, shifts] = getPotentiallyShiftedRegions(globalDomainLength, unshiftedRegion);
+			// Before every set of push_backs make sure there is enough space for this set + all remaining.
+			// Work with the assumption that the others are of the same size as the current ones.
+			// This is potentially an overestimate but avoids a large number of resizes.
```

Comment:
Is this assumption because the sizes of the regions decrease, so you can guarantee this is an estimate? Or is this simply a guess? In which case, what is the reasoning behind the guess? Ultimately, I don't really mind if the reasoning is something as simple as "better than doing nothing but otherwise arbitrary," but the reasoning behind the choice should be somewhat more documented.

Reply:
The idea was to call reserve for every region and reserve the vector as if all following regions are of this region's size. Thus it would always be an overestimate since in the end the vector is reserved according to the biggest region. Calls to reserve with smaller sizes don't do anything.

Looking at the code now I realize that it is not doing what I intended... Let me create a fix...

Reply:
have a look at the updated doc -> c317648